include/dlaf/communication/kernels/reduce.h

```diff
@@ -37,127 +34,179 @@ namespace comm {
+              matrix::internal::copy(tile_cpu, tile_gpu, args...);
+            },
+            std::move(tile_reduced));
+      })) |
+      start_detached();
 }
 
-template <class T, Device D, template <class> class Future>
+// TODO scheduleReduceSend with future will require to move the actual value, not the cref
+template <class T>
 void scheduleReduceSend(const comm::Executor& ex, comm::IndexT_MPI rank_root,
                         hpx::future<common::PromiseGuard<comm::Communicator>> pcomm, MPI_Op reduce_op,
-                        Future<matrix::Tile<T, D>> tile) {
-  using hpx::dataflow;
-  using hpx::unwrapping;
-
-  using common::internal::makeItContiguous_o;
-  using matrix::Tile;
-  using matrix::duplicateIfNeeded;
-  using internal::reduceSend_o;
-  using matrix::unwrapExtendTiles;
-
-  const auto& ex_copy = getHpExecutor<Backend::MC>();
+                        hpx::shared_future<matrix::Tile<const T, Device::CPU>> tile) {
+  using namespace hpx::execution::experimental;
+
+  return keep_future(std::move(tile)) |
+         let_value(hpx::unwrapping([ex, rank_root, reduce_op, pcomm = std::move(pcomm)](
+                                       const matrix::Tile<const T, Device::CPU>& tile) mutable {
+           return internal::senderReduceSend<T>(ex, rank_root, std::move(pcomm), reduce_op,
+                                                just(std::cref(tile)));
+         })) |
+         start_detached();
+}
 
+// TODO scheduleReduceSend with future will require to move the actual value, not the cref
+template <class T, Device D>
+void scheduleReduceSend(const comm::Executor& ex, comm::IndexT_MPI rank_root,
+                        hpx::future<common::PromiseGuard<comm::Communicator>> pcomm, MPI_Op reduce_op,
+                        hpx::shared_future<matrix::Tile<const T, D>> tile) {
   // Note:
   //
-  // TILE can be on CPU or GPU
+  // GPU ---> Duplicate ---> (cCPU --> MPI)
   //
-  // TILE_CPU ----> duplicateIfNeeded<CPU> ----> TILE_CPU (no-op)
-  // TILE_GPU ----> duplicateIfNeeded<CPU> ----> TILE_CPU
-  hpx::shared_future<Tile<const T, Device::CPU>> tile_cpu =
-      duplicateIfNeeded<Device::CPU>(std::move(tile));
+  // where: cCPU = contiguous CPU
 
-  // Note:
-  //
-  // TILE_CPU -+-> makeContiguous ---> CONT_BUF ---+--> mpi_call--+
-  //           |                                   |              |
-  //           +-----------> TILE_CPU -------------+--------------+
-  auto cont_buf = dataflow(ex_copy, unwrapping(makeItContiguous_o), tile_cpu);
+  using namespace hpx::execution::experimental;
+  using dlaf::internal::Policy;
+  using dlaf::matrix::internal::CopyBackend;
+
+  auto tile_cpu = dlaf::internal::transform(
+      Policy<CopyBackend<D, Device::CPU>::value>(hpx::threads::thread_priority::high),
+      [](const matrix::Tile<const T, Device::GPU>& tile_gpu, auto... args) {
+        return dlaf::matrix::Duplicate<Device::CPU>{}(tile_gpu, args...);
+      },
+      keep_future(std::move(tile)));
```

Comment:
This might've been intentional, but I just want to point out that this is not equivalent to `duplicateIfNeeded` since this will unconditionally duplicate the tile. You probably wanted this since CPU tiles won't typically be contiguous?

In general I wonder if these communication helpers would benefit from (another) helper sender adaptor to abstract away a bit of this complexity (not so much a comment about this PR but rather about the complexity I added earlier!). Something that looks a bit like `let_value`, but instead of giving you a reference to the input sender value, it gives you a reference to an already contiguous buffer:
```
with_contiguous_buffer(sender_of_tile, [](contiguous_buffer& buf) {
    return dataflow(mpi_exec, mpi_func, buf, pipeline(), ...);
    // or soon: return when_all(just(buf), pcomm, ...) | transform_mpi(mpi_func); 
});
```
So essentially what you've done now but refactored into separate functions. We can discuss this in more detail offline.

Reply:
> This might've been intentional, but I just want to point out that this is not equivalent to duplicateIfNeeded since this will unconditionally duplicate the tile. You probably wanted this since CPU tiles won't typically be contiguous?

Yes, exactly. I wanted to use directly `Duplicate`, because I'm sure it is needed. I thought it was a more clear solution.

> In general I wonder if these communication helpers would benefit from (another) helper sender adaptor to abstract away a bit of this complexity (not so much a comment about this PR but rather about the complexity I added earlier!). Something that looks a bit like let_value, but instead of giving you a reference to the input sender value, it gives you a reference to an already contiguous buffer:

I completely agree with you. It is something I want to do and I have in the back of my mind since https://github.com/eth-cscs/DLA-Future/pull/374#discussion_r636813607 😄

Reply:
Sounds good! I realize now that I never responded to your question on #374. I think you already know the answer but I'll respond anyway.

Reply:
For later: https://github.com/eth-cscs/DLA-Future/issues/485. Thanks @albestro for the issue!